src/catchup/CatchupRange.h

```diff
@@ -0,0 +1,77 @@
+#pragma once
+
+// Copyright 2017 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "ledger/CheckpointRange.h"
+
+namespace stellar
+{
+
+// Range of ledger checkpoint to catchup.
+// Contains stop at value which must be in range of last checkpoint to catchup.
+// In normal conditions stop at value is equal to last ledger in last
+// checkpoint in range, but for manual and command line catchup it can be
+// smaller than that.
+//
+// If apply buckets is true, this catchup requires downloading and applying
+// buckets for the first checkpoint and then downloading and applying
+// transactions.
+//
+// If apply buckets is false, this catchup requires downloading and applying
+// transactions for whole range of checkpoints.
+//
+// When (firstCheckpoint() == lastCheckpoint()) && applyBuckets() this is
+// equivalent to old CATCHUP_MIMIMAL.
+//
+// For old CATCHUP_COMPLETE applyBuckets() is always false, as CATCHUP_COMPLETE
+// only applies transactions.
+//
+// For old CATCHUP_RECENT value of applyBuckets() depends on value of last
+// closed ledger - if CATCHUP_RECENT would go before it, it behaves exactly as
+// CATCHUP_COMPLETE. If not, applyBuckets() is true and miminal catchup will
+// be done first.
+class CatchupRange
```

Comment:
This class doesn't seem to make sense to me. Especially as you are introducing a CatchupConfiguration in the next commit.

Reply:
CatchupRange life time is as follows:
* CatchupConfiguration is created with, LCL, catchup-to and catchup-count values
* ResolvedCatchupConfiguration is created that replaces catchup-to to a value that is not CatchupConfiguration::CURRENT (in case of command line catchups)
* CatchupRange is computed from ResolvedCatchupConfiguration and local state of ledger

So CatchupRange is internal class to CatchupWork as its creation requires many information that is just not available at the moment of CatchupConfiguration creation (mainly what is the last ledger on network).

Also CatchupConfiguration shows what we want to do (get last X ledgers to ledger Y or to last ledger) and CatchupRange shows how to do that (apply buckets at checkpoint F and then apply transactions to checkpoint L).

Reply:
What I was saying is that CatchupWork passed down to the various *Work doesn't make much sense and creates weird couplings: the old interfaces were only taking what the *Work needed, now they take this object type as if they cared about it, but they don't. For example `DownloadAndApplyTransactionsWork` doesn't care about `applyBuckets` or why the range is what it is, it just needs a start and an end.